src/transactions/TransactionUtils.cpp

```diff
@@ -1172,6 +1174,321 @@ removeOffersByAccountAndAsset(AbstractLedgerTxn& ltx, AccountID const& account,
+    if (tryGetEntrySponsor(trustLine, sponsor))
+    {
+        return sponsor;
+    }
+
+    return trustLine.current().data.trustLine().accountID;
+}
+
+static void
+prefetchForRevokeFromPoolShareTrustLines(
+    AbstractLedgerTxn& ltx, AccountID const& accountID,
+    std::vector<LedgerTxnEntry>& poolShareTrustLines)
+{
+    // first prefetch the liquidity pools and pool share trustline sponsors
+    UnorderedSet<LedgerKey> keys;
+    for (auto& trustLine : poolShareTrustLines)
+    {
+        keys.emplace(liquidityPoolKey(
+            trustLine.current().data.trustLine().asset.liquidityPoolID()));
+
+        AccountID sponsor;
+        if (tryGetEntrySponsor(trustLine, sponsor))
+        {
+            keys.emplace(accountKey(sponsor));
+        }
+    }
+    ltx.prefetch(keys);
+
+    // now prefetch the asset trustlines
+    keys.clear();
+    for (auto const& trustLine : poolShareTrustLines)
+    {
+        auto pool = loadLiquidityPool(
+            ltx, trustLine.current().data.trustLine().asset.liquidityPoolID());
+
+        auto params = [&]() -> auto&
+        {
+            return pool.current()
+                .data.liquidityPool()
+                .body.constantProduct()
+                .params;
+        };
+
+        if (params().assetA.type() != ASSET_TYPE_NATIVE &&
+            !isIssuer(accountID, params().assetA))
+        {
+            keys.emplace(trustlineKey(accountID, params().assetA));
+        }
+
+        if (params().assetB.type() != ASSET_TYPE_NATIVE &&
+            !isIssuer(accountID, params().assetB))
+        {
+            keys.emplace(trustlineKey(accountID, params().assetB));
+        }
+    }
+    ltx.prefetch(keys);
+}
+
+static ClaimableBalanceID
+getRevokeID(AccountID const& txSourceID, SequenceNumber txSeqNum,
+            uint32_t opIndex, PoolID const& poolID, Asset const& asset)
+{
+    OperationID operationID;
+    operationID.type(ENVELOPE_TYPE_POOL_REVOKE_OP_ID);
+    operationID.revokeID().sourceAccount = txSourceID;
+    operationID.revokeID().seqNum = txSeqNum;
+    operationID.revokeID().opNum = opIndex;
+    operationID.revokeID().liquidityPoolID = poolID;
+    operationID.revokeID().asset = asset;
+
+    ClaimableBalanceID balanceID;
+    balanceID.type(CLAIMABLE_BALANCE_ID_TYPE_FROM_POOL_REVOKE);
+    balanceID.fromPoolRevoke() = xdrSha256(operationID);
+
+    return balanceID;
+}
+
+static Claimant
+makeUnconditionalClaimant(AccountID const& accountID)
+{
+    Claimant c;
+    c.v0().destination = accountID;
+    c.v0().predicate.type(CLAIM_PREDICATE_UNCONDITIONAL);
+
+    return c;
+}
+
+bool
+removeOffersAndPoolShareTrustLines(AbstractLedgerTxn& ltx,
+                                   AccountID const& accountID,
+                                   Asset const& asset,
+                                   AccountID const& txSourceID,
+                                   SequenceNumber txSeqNum, uint32_t opIndex)
+{
+    removeOffersByAccountAndAsset(ltx, accountID, asset);
```

Comment:
If `ltxInner` is passed into `removeOffersByAccountAndAsset`, then `ltxInner.commit()` will need to be called whenever this function returns early (eg. if `poolShareTrustLines` is empty). I thought it would be safer this way in case another termination condition is added without the `commit`.